### PR TITLE
issue #156: allow unequal array lengths in sublabels

### PR DESCRIPTION
**Pull request by k-allagbe**
#156

### CODE REVIEW

File: datastore/db/bytebase/FertiScan/new_inspection_function.sql [hunk 1]
```diff
@@ -233,28 +236,36 @@ BEGIN
...
-					FALSE
-				);
-			END LOOP;
-		END IF;
-   END LOOP;    
+		en_values := COALESCE(input_json -> key_string -> 'en', '[]'::jsonb);
+		fr_values := COALESCE(input_json -> key_string -> 'fr', '[]'::jsonb);
+
+		-- Determine the maximum length of the arrays
+		max_length := GREATEST(
```
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/new_inspection_function.sql:247]: Do you think we should raise a notice if the length is not equal? This could be some information left in the comment section of the JSON between layers.

(this could be a future issue)

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py [hunk 2]
```diff
@@ -206,8 +207,84 @@ def test_update_sub_labels(self):
...
+            # Execute the function with mismatched sub labels
+            self.cursor.execute(
+                "SELECT update_sub_labels(%s, %s);",
+                (self.label_id, mismatched_sub_labels),
             )
+        except Exception as e:
+            self.fail(f"update_sub_labels raised an exception unexpectedly: {e}")
+
+        # Verify that no exception was raised and check the data integrity
+        self.cursor.execute(
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/db/update_inspection/test_update_sub_labels.py:241]: Can you use a 📦 sub_label.some_function instead of using manual queries?

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py [hunk 3]
```diff
@@ -206,8 +207,84 @@ def test_update_sub_labels(self):
...
+            "SELECT text_content_fr, text_content_en FROM sub_label WHERE label_id = %s;",
+            (self.label_id,),
+        )
+        saved_data = self.cursor.fetchall()
+
+        # Expected data should still be saved even though the arrays were mismatched
+        expected_data = [
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/db/update_inspection/test_update_sub_labels.py:248]: I personally think it is enough to compare the number of sub_labels, no need to directly compare the content

Comment (Francois-Werbrouck) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py:248]: Also creating a `expected_data` feels weird, you should refer to your `mismatched_sub_labels` if you want to directly compare the content

File: tests/fertiscan/db/update_inspection/test_upsert_inspection.py [hunk 4]
```diff
@@ -49,8 +50,8 @@ def tearDown(self):
     def test_insert_new_inspection_minimal_fields(self):
         # Insert a new inspection with minimal required fields
         self.cursor.execute(
-            'SELECT upsert_inspection(%s, %s, %s, %s, %s, %s);',
-            (None, self.label_info_id, self.inspector_id, None, None, False),
+            "SELECT upsert_inspection(%s, %s, %s, %s, %s, %s, %s);",
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/db/update_inspection/test_upsert_inspection.py:53]: I suggest using `sub_label.new_inpection()` when you want to create an inspection. 
Also, an inspection is never upserted in our workflow

File: tests/fertiscan/metadata/test_inspection.py [hunk 5]
```diff
@@ -186,6 +182,44 @@ def test_missing_sub_label(self):
         self.assertListEqual(data["instructions"]["fr"], [])
         self.assertListEqual(data["instructions"]["en"], [])
 
+    def test_unequal_sub_label_lengths_and_order(self):
+        self.analyse["instructions_en"] = ["one"]
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/metadata/test_inspection.py:186]: you should instantiate `expected_instructions_en` and then 

```suggestion
        self.analyse["instructions_en"] = expected_instructions_en
```

This way we can 100% be certain of the order on the insert

File: tests/fertiscan/metadata/test_inspection.py [hunk 6]
```diff
@@ -186,6 +182,44 @@ def test_missing_sub_label(self):
...
+        data = metadata.build_inspection_export(
+            self.cursor, str(inspection_id), label_information_id
+        )
+        data = json.loads(data)
+
+        self.assertIsNotNone(data["instructions"])
+        self.assertIsNotNone(data["instructions"]["en"])
+        self.assertIsNotNone(data["instructions"]["fr"])
+
+        expected_instructions_en = ["one", ""]
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/metadata/test_inspection.py:209]:
```suggestion

```

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py [hunk 7]
```diff
@@ -206,8 +207,84 @@ def test_update_sub_labels(self):
         for expected_updated_item in expected_updated_data:
             self.assertTrue(
                 expected_updated_item in updated_data,
-                f"Expected updated item {expected_updated_item} was not found in the updated data"
+                f"Expected updated item {expected_updated_item} was not found in the updated data",
+            )
+
+    def test_update_sub_labels_with_mismatched_arrays(self):
+        # Create sub labels with mismatched 'fr' and 'en' arrays
+        mismatched_sub_labels = json.dumps(
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/db/update_inspection/test_update_sub_labels.py:215]: I suggest you use the dict from your set up and remove one enrty from instructrion. This way makes the files lighter to read and easier to maintain

File: datastore/db/bytebase/FertiScan/new_inspection_function.sql (hunk 1 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/new_inspection_function.sql:247]: Done.

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py (hunk 2 above)
Comment (k-allagbe) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: Done. The rest will be properly done in https://github.com/ai-cfia/ailab-datastore/issues/158

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py [hunk 8]
```diff
@@ -206,8 +207,84 @@ def test_update_sub_labels(self):
...
+
+        # Expected data should still be saved even though the arrays were mismatched
+        expected_data = [
+            ("1. Dissoudre 50g dans 10L d'eau.", "1. Dissolve 50g in 10L of water."),
+            ("2. Appliquer toutes les 2 semaines.", "2. Apply every 2 weeks."),
+            ("", "3. Store in a cool, dry place."),
+        ]
+        self.assertEqual(
+            len(saved_data),
+            3,
```
[the remark this answers is not in the document]
Comment (k-allagbe) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: Done. I now use the original data variable.

Note that because this:
```sql
    CREATE TABLE "fertiscan_0.0.14"."sub_label" (
        "id" uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
        "text_content_fr" text NOT NULL DEFAULT '',  <--- not null here
        "text_content_en" text NOT NULL DEFAULT '',  <--- and here
        "label_id" uuid NOT NULL REFERENCES "fertiscan_0.0.14"."label_information" ("id") ON DELETE CASCADE,
        "edited" boolean, --this is because with the current upsert we can not determine if it was edited or not
        "sub_type_id" uuid NOT NULL REFERENCES "fertiscan_0.0.14"."sub_type" ("id")
    );

```
I resorted to use an empty string for the missing sub label strings. Which is why in the test, `en` and `fr` have equal lengths in the output. Let me know if this is satisfactory.

File: tests/fertiscan/db/update_inspection/test_upsert_inspection.py (hunk 4 above)
Comment (k-allagbe) [reply tests/fertiscan/db/update_inspection/test_upsert_inspection.py]: These will be fixed in https://github.com/ai-cfia/ailab-datastore/issues/158.

File: tests/fertiscan/metadata/test_inspection.py (hunk 5 above)
Comment (k-allagbe) [reply tests/fertiscan/metadata/test_inspection.py]: Sets allow handling such situations where the order might change: 
```python
        self.assertEqual(
            set(data["instructions"]["en"]),
            set(expected_instructions_en),
            "Instructions EN mismatch",
        )
```

I made the modification nonetheless.

File: tests/fertiscan/metadata/test_inspection.py (hunk 6 above)
Comment (k-allagbe) [reply tests/fertiscan/metadata/test_inspection.py]: Done

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py (hunk 7 above)
Comment (k-allagbe) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: Done.

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Francois-Werbrouck) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: An empty string is still a value that will be registered and does not represent the Edge case we are attempting to solve.

If you want to really tests the length difference I suggest using an array of different length and modify the code to handle such a case. The getter should return an the fr and en with not the name length.

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql [hunk 9]
```diff
@@ -1,33 +1,30 @@
...
-    fr text[] ;
-    en text[] ;
 BEGIN
     SELECT jsonb_object_agg(
                 sub_type.type_en, 
                 jsonb_build_object(
-				    'en', COALESCE(sub_label_data.texts_en, array[]::text[]),
-				    'fr', COALESCE(sub_label_data.texts_fr, array[]::text[])
-				)			
+				    'en', COALESCE(NULLIF(sub_label_data.texts_en, array[]::text[]), NULL),
```
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql:11]: This doesn't work with our standard for the getters. It must return an empty array if there are no entry for this object. I dont think you need to change the getter

File: datastore/db/bytebase/FertiScan/update_inspection_function.sql [hunk 10]
```diff
@@ -375,6 +375,30 @@ END;
...
 
 
+-- Function to check if both text_content_fr and text_content_en are NULL, and skip insertion if true
+CREATE OR REPLACE FUNCTION "fertiscan_0.0.14".check_null_sub_label()
+RETURNS TRIGGER AS $$
+BEGIN
+    -- Check if both text_content_fr and text_content_en are NULL
+    IF NEW.text_content_fr IS NULL AND NEW.text_content_en IS NULL THEN
+        -- Raise a notice and skip the insertion by returning NULL
+        RAISE NOTICE 'Skipping insertion because both text_content_fr and text_content_en are NULL';
```
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/update_inspection_function.sql:385]: I'd even make this an Exception, to make sure it doesn't happen. 
```suggestion
        RAISE Exception 'Skipping insertion because both text_content_fr and text_content_en are NULL';
```

File: tests/fertiscan/db/test_sub_label.py [hunk 11]
```diff
@@ -218,3 +218,38 @@ def test_get_all_sub_label(self):
...
+        )
+
+        # Fetch sub-label data as JSON
+        sub_label_data = sub_label.get_sub_label_json(self.cursor, self.label_id)
+
+        # Assert that the 'en' value for the first sub-label is not empty
+        self.assertTrue(sub_label_data[self.type_en]['en'], "Expected 'en' field to have content")
+        
+        # Assert that the 'fr' value for the first sub-label is empty (since it was NULL)
+        self.assertFalse(sub_label_data[self.type_en]['fr'], "Expected 'fr' field to be empty")
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/db/test_sub_label.py:249]: We should verify the length of the array to be 0

**Review by Francois-Werbrouck — CHANGES_REQUESTED** [review]

**Comment by Francois-Werbrouck** [comment]
Here is a snippet of code on how we verify the empty inspection. Take a look at the empty list to understand how we expect the structure to be like if empty. This is important so the FE doesn't receive Null value where they do not expect.

Snippet from 📦 tests.fertiscan.test_datastore
```python
def loop_into_empty_dict(dict_data):
    passing = True
    for key, value in dict_data.items():
        if isinstance(value, dict):
            loop_into_empty_dict(value)
        elif isinstance(value, list):
            if len(value) == 0:
                continue
            else:
                passing = False
        elif isinstance(value, bool):
            if not value:
                continue
            else:
                passing = False
        else:
            if value is None:
                continue
            elif validator.is_valid_uuid(value):
                continue
            else:
                passing = False
    return passing
```

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql (hunk 9 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql]: Yes, that was not intended. Sometimes gpt gets a bit overzealous. Fixed.

File: datastore/db/bytebase/FertiScan/update_inspection_function.sql (hunk 10 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/update_inspection_function.sql]: Done

File: tests/fertiscan/db/test_sub_label.py (hunk 11 above)
Comment (k-allagbe) [reply tests/fertiscan/db/test_sub_label.py]: It's all the same, but done.

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (k-allagbe) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: How about now?

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql [hunk 12]
```diff
@@ -1,33 +1,30 @@
...
            )
     INTO result_json
     FROM sub_type
-    left JOIN (
+    LEFT JOIN (
         SELECT 
             sub_type_id,
-            array_agg(sub_label.text_content_en ORDER BY sub_label.id) AS texts_en,
-            array_agg(sub_label.text_content_fr ORDER BY sub_label.id) AS texts_fr
+            array_agg(sub_label.text_content_en ORDER BY sub_label.id) FILTER (WHERE sub_label.text_content_en IS NOT NULL) AS texts_en,
```
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql:20]: Wont this produce problem (not returning rows) if the content_en is null but not the content_fr?

File: datastore/db/bytebase/FertiScan/update_inspection_function.sql [hunk 13]
```diff
@@ -416,8 +440,8 @@ BEGIN
         FOR i IN 0..(max_length - 1)
         LOOP
             -- Extract values or set to empty string if not present
-            fr_value := COALESCE(fr_values->>i, '');
-            en_value := COALESCE(en_values->>i, '');
+            fr_value := fr_values->>i;
```
Comment (Francois-Werbrouck) [inline datastore/db/bytebase/FertiScan/update_inspection_function.sql:443]: How do we manage the edge case where both are Null? (also in the new_inspection)

**Comment by Francois-Werbrouck** [comment]
I see we modified `new_inspection`, however, I dont see a new test showcasing it can handle different type of Null combination

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql (hunk 12 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql:20]: I used the filter to avoid such arrays:
```json
{"en": ["label", null], "fr": ["label", "label"]}
```
Instead it will be:
```json
{"en": ["label"], "fr": ["label", "label"]}
```

The code in the jsonb_build_object seems to only avoid:
```json
{"en": null, "fr": ["label", "label"]}
```
Instead it will be:
```json
{"en": [], "fr": ["label", "label"]}
```

Incorrect?

File: datastore/db/bytebase/FertiScan/update_inspection_function.sql (hunk 13 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/update_inspection_function.sql:443]: With a trigger. Please check file `datastore/db/bytebase/FertiScan/update_inspection_function.sql`

```sql
-- Function to check if both text_content_fr and text_content_en are NULL, and skip insertion if true
CREATE OR REPLACE FUNCTION "fertiscan_0.0.14".check_null_sub_label()
RETURNS TRIGGER AS $$
BEGIN
    -- Check if both text_content_fr and text_content_en are NULL
    IF NEW.text_content_fr IS NULL AND NEW.text_content_en IS NULL THEN
        -- Raise a notice and skip the insertion by returning NULL
        RAISE Exception 'Skipping insertion because both text_content_fr and text_content_en are NULL';
        RETURN NULL;
    END IF;
    -- Allow the insert if at least one is not NULL
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;


-- Trigger to call check_null_sub_label() before inserting into sub_label
DROP TRIGGER IF EXISTS before_insert_sub_label ON "fertiscan_0.0.14".sub_label;
CREATE TRIGGER before_insert_sub_label
BEFORE INSERT ON "fertiscan_0.0.14".sub_label
FOR EACH ROW
EXECUTE FUNCTION check_null_sub_label();
```

In that case, we raise an exception as we agreed.

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql (hunk 12 above)
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql:20]: While reading the code I was under the impression that if the following exists:

| id|text_en | text_fr |
| --|------ | ------- |
| 1|text | texte |
| 2|text | Null |
| 3|Null | texte |
| 4 (should not exists in theory) | Null | Null |

the filter would only return the row with rowID=1 by blocking a row if it has a null? 

If the filter still returns all the rows with rowID= [1,2,3] this sits good with me.

I also feel it is important to return a null because if the following exists for an inspection:

| id|text_en | text_fr |
| --|------ | ------- |
| 1 | text1 | texte1 |
| 2 | text2 | Null |
| 3 | text3| texte3 |

the arrays would look like

```{"en": [text1,text2,text3], "fr": ["texte1", "text3"]}``` 

If returned for an update (with no modification) then the db will be updated as such:

| id|text_en | text_fr |
| --|------ | ------- |
| 1 | text1 | texte1 |
| 2 | text2 | texte3 |
| 3 | text3| ??? |

this is if the db doesnt throw an exception since we are looping both array until we reach the greatest index. In this case the biggest index is 3 and the fr array has 2 values. I'm not so sure about this working, ence why I'd like to see those tests cases tested with the whole pipeline 

>I see we modified new_inspection, however, I dont see a new test showcasing it can handle different type of Null combination

maybe also add in a test with the specified edge case that I mentionned with new_inspection and update_inspection to make sure the whole process behave as intented.

We might sit with the FE to make sure they can handle Null value. 

Otherwise a solution I see rolling back to `Default ''` in the schema, this might solve the edge case if there is problems

File: datastore/db/bytebase/FertiScan/update_inspection_function.sql (hunk 13 above)
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/update_inspection_function.sql:443]: Hmm, I would of checked for a double null value even before inserting to keep the operational cost as minimal as possible. Adding a trigger makes it more complex but create a safety net even outside the function, this solution works and I forgot we agreed upon it.

Thank you for pointing it out to me, This comment can be resolved.

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql (hunk 12 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql:20]: This:
```json
{"en": ["text1","text2","text3"], "fr": ["texte1", "text3"]}
```
was per your requirements.

`fr_value := fr_values->>i;` returns null for an overflow, not an exception.

I'm fine with returning a placeholder, because it preserves the correct positions between en and fr. I suggest the placeholder be an empty string instead of null however. Indeed, in the frontend we have this:
```typescript
export const populateFromJSON = (form: Data, data: Inspection) => {
  form.sections.forEach((section) => {
    section.inputs.forEach((input) => {
      const value = access_from_str(data, input.id);
      if (typeof value == "string") {
        input.value = [value];
      } else if (Array.isArray(value)) {
        if (value.length == 0) {
          input.isAlreadyTable = true;
          input.value = [""];
        } else if (typeof value[0] == "string") {  //<---------------- here
          input.value = value;
          input.isAlreadyTable = true;
        } else if (typeof value[0] == "object") {   //<---------------- and here
          input.value = value;
          input.isInputObjectList = true;
        }
      } else if (typeof value == "object") {
        // eslint-disable-next-line
        // @ts-ignore
        input.value = [value];
      }
    });
  });
  // for state update the function must return a new object
  return form.copy();
};
```

If ever the first checked value is null, the rest will be ignored.

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Francois-Werbrouck) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: With a further understanding of the issue, the following is not true:
>The getter should return an the fr and en with not the name length.
The array should be returned of same length, however, we should not save an empty value. Unless we cannot solve this without inserting an empty string

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql (hunk 12 above)
Comment (Francois-Werbrouck) [reply datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql:20]: >fr_value := fr_values->>i; returns null for an overflow, not an exception.

So the db will be 

id | text_en | text_fr
-- | -- | --
1 | text1 | texte1
2 | text2 | texte3
3 | text3 | Null

Instead of being

id | text_en | text_fr
-- | -- | --
1 | text1 | texte1
2 | text2 | Null
3 | text3 | texte3

>was per your requirements.

...?
I think you are referring to
>An empty string is still a value that will be registered and does not represent the Edge case we are attempting to solve.

What I mean is, we do not want to save an empty string in the database when there are no record of a value being there.
>I'm fine with returning a placeholder, because it preserves the correct positions between en and fr. I suggest the placeholder be an empty string instead of null

I'm also ok with returning an empty string as a place holder. This could also be used by the frontend to detect empty fields that should be flagged to the inspector. However, We must check, once we receive it back that the update is performed correctly since the update is actually a delete + insert .


- Case 1: 
 -> The place holder returned as some text (length>0)
l---> New.text_content = some text
- Case 2: 
-> The returned value of a sub_label_string is an empty string (length=0)
l---> New.text_content = Null

File: tests/fertiscan/db/update_inspection/test_update_sub_labels.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (k-allagbe) [reply tests/fertiscan/db/update_inspection/test_update_sub_labels.py]: Same length, but with empty strings as placeholders.

File: datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql (hunk 12 above)
Comment (k-allagbe) [reply datastore/db/bytebase/FertiScan/get_inspection/get_sub_label.sql]: Not sure I understand the last part. I've already updated the update_sub_labels function and tested it.

File: tests/fertiscan/metadata/test_inspection.py [hunk 14]
```diff
@@ -370,3 +370,144 @@ def test_no_empty_specifications(self):
...
+        inspection_data = json.loads(inspection_data)
+
+        # Assert that null in cautions_en is replaced by an empty string
+        self.assertIsNotNone(inspection_data["cautions"])
+        self.assertEqual(
+            inspection_data["cautions"]["en"], ["Warning 1", "", "Warning 2"]
+        )
+        self.assertEqual(
+            inspection_data["cautions"]["fr"],
+            ["Avertissement 1", "Avertissement 2", "Avertissement 3"],
```
Comment (Francois-Werbrouck) [inline tests/fertiscan/metadata/test_inspection.py:436]: This helps makes the tests more maintenable 

```suggestion
            formatted_analysis_dict["cautions"]["fr"],
```

**Review by Francois-Werbrouck — APPROVED** [review]
All the tests are there and passing. This looks good to me. 

You could include my last comment in your #158

**Comment by Francois-Werbrouck** [comment]
>You could include my last comment in your https://github.com/ai-cfia/ailab-datastore/issues/158

nvm you've already done it 🥇
